R/test-server.R

```diff
@@ -50,20 +50,21 @@ isModuleServer <- function(x) {
 #'   # Any additional arguments, below, are passed along to the module.
 #' })
 #' @export
-testServer <- function(app = NULL, expr, args = list()) {
+testServer <- function(app = NULL, expr, args = list(), session = makeMockSession()) {
 
   require(shiny)
 
   quosure <- rlang::enquo(expr)
-  session <- getDefaultReactiveDomain()
 
-  if (inherits(session, "MockShinySession"))
-    stop("Test expressions may not call testServer()")
-  if (inherits(session, "session_proxy")
-      && inherits(get("parent", envir = session), "MockShinySession"))
-    stop("Modules may not call testServer()")
+  local({
```

Comment:
If `session` now defaults to `makeMockSession()`, what is the purpose of moving this code into local and checking the `getDefaultReactiveDomain()`?

Reply:
The idea for exposing the `session` argument was that expert users might want to subclass/override parts of a MockShinySession to do stuff like assign to `user` or `groups` to test auth, or to provide bogus clientData.

The check endures because the average user should still be aware that nested testServer is bad.